File: platform/player-manager.js
// Bucu Core - Player Manager (JavaScript)
// Manages player objects and lifecycle

const PlayerManager = {
    _players: new Map(),  // Map<number, Player>
};

// Player class
class Player {
    constructor(source, name, identifier) {
        this.id = source;
        this.name = name || "Unknown";
        this.identifier = identifier || "";
        this.ping = 0;
    }
    
    // Get player permission
    getPermission() {
        if (typeof PermissionManager !== 'undefined') {
            return PermissionManager.getPermission(this.id);
        }
        return "user";
    }
    
    // Set player permission
    setPermission(role) {
        if (typeof PermissionManager !== 'undefined') {
            return PermissionManager.setPermission(this.id, role);
        }
        return false;
    }
    
    // Get player metadata
    getMeta(key) {
        if (typeof StateCache !== 'undefined') {
            const cacheKey = `player:${this.id}:meta:${key}`;
            return StateCache.get(cacheKey);
        }
        return null;
    }
    
    // Set player metadata
    setMeta(key, value) {
        if (typeof StateCache !== 'undefined') {
            const cacheKey = `player:${this.id}:meta:${key}`;
            return StateCache.set(cacheKey, value);
        }
        return false;
    }
    
    // Kick player
    kick(reason = "Kicked by administrator") {
        if (typeof Logger !== 'undefined') {
            Logger.info(`Kicking player ${this.name} (${this.id}): ${reason}`);
        }
        
        // FiveM native call (will be handled by adapter)
        if (typeof DropPlayer !== 'undefined') {
            DropPlayer(this.id.toString(), reason);
        }
        
        return true;
    }
    
    // Update player ping
    updatePing() {
        if (typeof GetPlayerPing !== 'undefined') {
            this.ping = GetPlayerPing(this.id.toString());
        }
    }
    
    // Get player identifiers (license, steam, etc.)
    getIdentifiers() {
        if (typeof GetPlayerIdentifiers !== 'undefined') {
            return GetPlayerIdentifiers(this.id.toString());
        }
        return [];
    }
    
    // Check if player is online
    isOnline() {
        return PlayerManager.getPlayer(this.id) !== null;
    }
}

// PlayerManager methods

// Create and register a player
PlayerManager.createPlayer = function(source, name, identifier) {
    const player = new Player(source, name, identifier);
    this._players.set(source, player);
    
    if (typeof Logger !== 'undefined') {
        Logger.info(`Player created: ${name} (${source}) [${identifier}]`);
    }
    
    // Emit player connected event
    if (typeof EventSystem !== 'undefined') {
        EventSystem.emit("player:connected", player);
    }
    
    return player;
};

// Get a player by source
PlayerManager.getPlayer = function(source) {
    return this._players.get(source) || null;
};

// Get all players
PlayerManager.getPlayers = function() {
    return Array.from(this._players.values());
};

// Get player count
PlayerManager.getPlayerCount = function() {
    return this._players.size;
};

// Remove a player
PlayerManager.removePlayer = function(source) {
    const player = this._players.get(source);
    
    if (!player) {
        return false;
    }
    
    if (typeof Logger !== 'undefined') {
        Logger.info(`Player removed: ${player.name} (${source})`);
    }
    
    // Emit player disconnected event
    if (typeof EventSystem !== 'undefined') {
        EventSystem.emit("player:disconnected", player);
    }
    
    // Clear player metadata from cache
    if (typeof StateCache !== 'undefined') {
        const metaKeys = StateCache.keys();
        const prefix = `player:${source}:meta:`;
        
        metaKeys.forEach(key => {
            if (key.startsWith(prefix)) {
                StateCache.delete(key);
            }
        });
    }
    
    // Clear permission cache
    if (typeof PermissionManager !== 'undefined') {
        PermissionManager.clearPermission(source);
    }
    
    // Remove from players map
    this._players.delete(source);
    
    return true;
};

// Get player by identifier
PlayerManager.getPlayerByIdentifier = function(identifier) {
    for (const player of this._players.values()) {
        if (player.identifier === identifier) {
            return player;
        }
    }
    return null;
};

// Get player by name (partial match)
PlayerManager.getPlayerByName = function(name) {
    const searchName = name.toLowerCase();
    
    for (const player of this._players.values()) {
        if (player.name.toLowerCase().includes(searchName)) {
            return player;
        }
    }
    
    return null;
};

// Update all player pings
PlayerManager.updateAllPings = function() {
    for (const player of this._players.values()) {
        player.updatePing();
    }
};

// Export for Node.js and FiveM
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { PlayerManager, Player };
}

// Make available globally in FiveM
if (typeof global !== 'undefined') {
    global.PlayerManager = PlayerManager;
    global.Player = Player;
}

File: platform/player-manager.lua
-- Bucu Core - Player Manager
-- Manages player objects and lifecycle

local PlayerManager = {
    _players = {},  -- { [source] = Player }
}

-- Player class
local Player = {}
Player.__index = Player

-- Create a new player object
function Player.new(source, name, identifier)
    local self = setmetatable({}, Player)
    
    self.id = source
    self.name = name or "Unknown"
    self.identifier = identifier or ""
    self.ping = 0
    
    return self
end

-- Get player permission
function Player:getPermission()
    if PermissionManager then
        return PermissionManager:getPermission(self.id)
    end
    return "user"
end

-- Set player permission
function Player:setPermission(role)
    if PermissionManager then
        return PermissionManager:setPermission(self.id, role)
    end
    return false
end

-- Get player metadata
function Player:getMeta(key)
    if StateCache then
        local cacheKey = "player:" .. self.id .. ":meta:" .. key
        return StateCache:get(cacheKey)
    end
    return nil
end

-- Set player metadata
function Player:setMeta(key, value)
    if StateCache then
        local cacheKey = "player:" .. self.id .. ":meta:" .. key
        return StateCache:set(cacheKey, value)
    end
    return false
end

-- Kick player
function Player:kick(reason)
    reason = reason or "Kicked by administrator"
    
    if Logger then
        Logger:info(string.format(
            "Kicking player %s (%d): %s",
            self.name,
            self.id,
            reason
        ))
    end
    
    -- FiveM native call (will be handled by adapter)
    if DropPlayer then
        DropPlayer(self.id, reason)
    end
    
    return true
end

-- Update player ping
function Player:updatePing()
    if GetPlayerPing then
        self.ping = GetPlayerPing(self.id)
    end
end

-- Get player identifiers (license, steam, etc.)
function Player:getIdentifiers()
    if GetPlayerIdentifiers then
        return GetPlayerIdentifiers(self.id)
    end
    return {}
end

-- Check if player is online
function Player:isOnline()
    return PlayerManager:getPlayer(self.id) ~= nil
end

-- PlayerManager methods

-- Create and register a player
function PlayerManager:createPlayer(source, name, identifier)
    local player = Player.new(source, name, identifier)
    self._players[source] = player
    
    if Logger then
        Logger:info(string.format(
            "Player created: %s (%d) [%s]",
            name,
            source,
            identifier
        ))
    end
    
    -- Emit player connected event
    if EventSystem then
        EventSystem:emit("player:connected", player)
    end
    
    return player
end

-- Get a player by source
function PlayerManager:getPlayer(source)
    return self._players[source]
end

-- Get all players
function PlayerManager:getPlayers()
    local players = {}
    for _, player in pairs(self._players) do
        table.insert(players, player)
    end
    return players
end

-- Get player count
function PlayerManager:getPlayerCount()
    local count = 0
    for _ in pairs(self._players) do
        count = count + 1
    end
    return count
end

-- Remove a player
function PlayerManager:removePlayer(source)
    local player = self._players[source]
    
    if not player then
        return false
    end
    
    if Logger then
        Logger:info(string.format(
            "Player removed: %s (%d)",
            player.name,
            source
        ))
    end
    
    -- Emit player disconnected event
    if EventSystem then
        EventSystem:emit("player:disconnected", player)
    end
    
    -- Clear player metadata from cache
    if StateCache then
        local metaKeys = StateCache:keys()
        local prefix = "player:" .. source .. ":meta:"
        
        for _, key in ipairs(metaKeys) do
            if key:sub(1, #prefix) == prefix then
                StateCache:delete(key)
            end
        end
    end
    
    -- Clear permission cache
    if PermissionManager then
        PermissionManager:clearPermission(source)
    end
    
    -- Remove from players table
    self._players[source] = nil
    
    return true
end

-- Get player by identifier
function PlayerManager:getPlayerByIdentifier(identifier)
    for _, player in pairs(self._players) do
        if player.identifier == identifier then
            return player
        end
    end
    return nil
end

-- Get player by name (partial match)
function PlayerManager:getPlayerByName(name)
    name = name:lower()
    
    for _, player in pairs(self._players) do
        if player.name:lower():find(name, 1, true) then
            return player
        end
    end
    
    return nil
end

-- Update all player pings
function PlayerManager:updateAllPings()
    for _, player in pairs(self._players) do
        player:updatePing()
    end
end

return PlayerManager
